pkg/plugins/policies/meshtrace/plugin/v1alpha1/plugin.go

```diff
@@ -20,11 +29,137 @@ func NewPlugin() core_plugins.Plugin {
+func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resources xds_context.Resources) (xds.TypedMatchingPolicies, error) {
 	return matchers.MatchedPolicies(api.MeshTraceType, dataplane, resources)
 }
 
-func (p plugin) Apply(rs *core_xds.ResourceSet, ctx xds_context.Context, proxy *core_xds.Proxy) error {
-	log.Info("apply is not implemented")
+func (p plugin) Apply(rs *xds.ResourceSet, ctx xds_context.Context, proxy *xds.Proxy) error {
+	policies, ok := proxy.Policies.Dynamic[api.MeshTraceType]
+	if !ok {
+		return nil
+	}
+	if len(policies.SingleItemRules.Rules) == 0 {
+		return nil
+	}
+
+	listeners := utils.GatherListeners(rs)
+	if err := applyToInbounds(policies.SingleItemRules, listeners.Inbound, proxy.Dataplane); err != nil {
+		return err
+	}
+	if err := applyToOutbounds(policies.SingleItemRules, listeners.Outbound, proxy.Dataplane); err != nil {
+		return err
+	}
+	if err := applyToClusters(policies.SingleItemRules, rs, proxy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func applyToInbounds(rules xds.SingleItemRules, inboundListeners map[xds.InboundListener]*envoy_listener.Listener, dataplane *core_mesh.DataplaneResource) error {
+	for _, inboundListener := range inboundListeners {
+		if err := configureListener(rules, dataplane, inboundListener, ""); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
+
+func applyToOutbounds(rules xds.SingleItemRules, outboundListeners map[mesh_proto.OutboundInterface]*envoy_listener.Listener, dataplane *core_mesh.DataplaneResource) error {
+	for _, outbound := range dataplane.Spec.Networking.GetOutbound() {
+		oface := dataplane.Spec.Networking.ToOutboundInterface(outbound)
+
+		listener, ok := outboundListeners[oface]
+		if !ok {
+			continue
+		}
+
+		serviceName := outbound.GetTagsIncludingLegacy()[mesh_proto.ServiceTag]
+
+		if err := configureListener(rules, dataplane, listener, serviceName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func configureListener(rules xds.SingleItemRules, dataplane *core_mesh.DataplaneResource, listener *envoy_listener.Listener, destination string) error {
+	serviceName := dataplane.Spec.GetIdentifyingService()
+	rawConf := rules.Rules[0].Conf
+	conf := rawConf.(*api.MeshTrace_Conf)
+
+	configurer := plugin_xds.Configurer{
+		Conf:             conf,
+		Service:          serviceName,
+		TrafficDirection: listener.TrafficDirection,
+		Destination:      destination,
+	}
+
+	for _, chain := range listener.FilterChains {
+		if err := configurer.Configure(chain); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func applyToClusters(rules xds.SingleItemRules, rs *xds.ResourceSet, proxy *xds.Proxy) error {
+	rawConf := rules.Rules[0].Conf
+
+	conf := rawConf.(*api.MeshTrace_Conf)
+
+	backend := conf.GetBackends()[0]
+	if backend == nil {
+		return nil
+	}
+
+	var endpoint *xds.Endpoint
+	var provider string
+
+	if backend.GetZipkin() != nil {
+		endpoint = endpointForZipkin(backend.GetZipkin())
+		provider = "zipkin"
```

Comment:
nit: I'd introduce a constant, probably in `api/v1alpha1` to show that this provider name has to be the same as in `xds/configurer.go`

Reply:
Created a constant but in `plugin/xds` because that is where the provider is created.